src/Common/HashTable/FixedHashTable.h

```diff
@@ -298,9 +303,14 @@ class FixedHashTable : private boost::noncopyable, protected Allocator, protecte
             return end();
 
         const Cell * ptr = buf;
-        auto buf_end = buf + NUM_CELLS;
-        while (ptr < buf_end && ptr->isZero(*this))
-            ++ptr;
+        if (!use_min_max_optimization())
```

Comment:
I think it also makes sense to abstract 

``` cpp
return use_min_max_optimization() ? but + min : buf;
```

and the same for `end` into a separate function. so all other code won't know anything inside this logic, it will just call `firstPopulatedCell()` .

Reply:
That's great. Thanks for your kind suggestion. The latest commit has packed the related code into `firstPopulatedCell()` and `lastPopulatedCell()`.